### ISSUE TITLE
Categories

### ISSUE THREAD

**Issue by davidwessman** [fsek/web#253]
We need some categorization on our little application.

For example the old image gallery had one model for `album_category` and one model for images called `subcategory`.

They basically do the same thing.
Could this be done with a Polymorphic `Category` instead? 

Thoughts? Ideas?

**Comment by jforberg** [comment]
I don't understand what you're talking about.

**Comment by davidwessman** [comment]
When we add an event, an album or something similar - I would like the ability to add a Category.

For example during initiation `nollning` could be used for a lot of them and it would be cool if you could look at a Category and see all the underlying joins.

Have anyone used `Act_as_taggable` before?

**Comment by jforberg** [comment]
Yes.

A.a.t is ok but it's also very easy (and much more flexible) to just write out the joins yourself.

The question to be asking here is do you really want global categories? It means that a lot of different code has to share logic. Note that the model code is exceedingly simple, just a plain SQL join and not much more.

The hard part is the views. Will you be able to share that code in a nice way?

**Comment by davidwessman** [comment]
How would you do the joins?

Will you add a field for the class and how can you add associations?
What do you think about polymorphic one?

Not sure about how much view there would be.
Perhaps a sidebar where a recently updated category can be featured or something.
Not 100% how to do the modelling.

For example a gallery, we have albums which could use a category (or many), then the images could have categories - for example "parbilder" could be separated.
Can we do this in a nice way with the same model?

**Comment by jforberg** [comment]
Well but do you want global categories? Do you want "parbilder" to be a general tag that can be used for news, events etc.

If not then you could have a concern that is "taggable". Then they could still be separate tables for different model types.

If you want tags to be global then you would need to use polymorphic assoications which look up the model class by a string column.

View: I mean the view where you assign categories, can be a bit tricky to do well.

**Comment by davidwessman** [comment]
Yeah, some will be very focused.

But implementing categories for every different kind? 

I'm thinking that you give categories a type:
`type: :general, title: :nollning`
`type: :image, title: :couple`

For model `:album` we can allow multiple categories, and allow the types: `:general, :album`
For model `:image` we only allow type: `:image`. 

Should we do this as different models?

**Comment by jforberg** [comment]
Ok, then I understand.

Remember, concerns apply the same logic to different models. So you don't need to reimplement anything.

I think a global categories table could be the right solution then. I don't think you want to have different "kinds" of tags though. If you make them global, you make them global. Having different types just adds extra special cases and breaks the user's mental model for no good reason.

Believe me, you do not want to have checkboxes or select in the long run. It will be unmanageable when many categories are added. Which they will be.

What you probably want is a fulltext search field with either Ajax or embedded Json feeding it. Will be an interesting exercise to implement I think.

Go to stack overflow and try adding a question with category. That is the UI we want.

**Comment by jforberg** [comment]
Something like this would work:

```sql
create table tags(
    id integer primary_key auto_increment,
    varchar name, -- remember to add an index on this column
    varchar model_class, -- or whatever rails calls it
    integer record_id, -- or whatever rails calls it, remember to index
)
```

**Comment by davidwessman** [comment]
Implemented in #368
